azext_edge/edge/providers/orchestration/resources/connector/opcua/certs.py

```diff
@@ -528,21 +540,156 @@ def _check_secret_name(
 
         return new_secret_name
 
-    def _upload_to_key_vault(self, keyvault_name: str, secret_name: str, file_path: str, cert_extension: str):
+    def _calculate_secret_expiration(
+        self,
+        file_path: str,
+        cert_extension: str,
+        expiration_date: Optional[str] = None,
+    ) -> datetime:
+        """
+        Calculate the appropriate expiration date for a Key Vault secret.
+
+        Priority order:
+        1. If expiration_date is provided by user, use that
+        2. For certificate files (.der, .crt), extract the certificate's expiration date
+        3. Default to 180 days from now
+
+        Args:
+            file_path: Path to the certificate/key file
+            cert_extension: File extension indicating the file type
+            expiration_date: Optional expiration date string in ISO 8601 format (user-provided)
+
+        Returns:
+            datetime: UTC-aware datetime for secret expiration
+        """
+        from datetime import timedelta
+        from dateutil import parser as date_parser
+
+        # Priority 1: User-provided expiration date
+        if expiration_date is not None:
+            try:
+                user_expiration = date_parser.isoparse(expiration_date)
+                # Ensure it's UTC-aware
+                if user_expiration.tzinfo is None:
+                    user_expiration = user_expiration.replace(tzinfo=timezone.utc)
+
+                # Validate it's in the future
+                if user_expiration <= datetime.now(timezone.utc):
+                    raise InvalidArgumentValueError(
+                        f"Expiration date must be in the future. Provided: {expiration_date}"
+                    )
+
+                logger.info(
+                    f"Using user-provided expiration date: {user_expiration.strftime('%Y-%m-%d %H:%M:%S UTC')} "
+                    f"for Key Vault secret."
+                )
+                return user_expiration
+            except (ValueError, TypeError) as e:
+                raise InvalidArgumentValueError(
+                    f"Invalid expiration date format. Expected ISO 8601 format (e.g., 2025-12-31T23:59:59Z). "
+                    f"Provided: {expiration_date}. Error: {str(e)}"
+                )
+
+        # Priority 2: For certificate files, try to extract the certificate's actual expiration date
+        if cert_extension in {X509FileExtension.DER.value, X509FileExtension.CRT.value}:
```

Comment:
we are also extracting expiry date from certificate in _process_cert_content, should we build a helper function so both can be used to reduce some duplication? thanks~